bgpd/bgpd.c

```diff
@@ -278,17 +282,38 @@ void bgp_router_id_zebra_bump(vrf_id_t vrf_id, const struct prefix *router_id)
 			if (bgp->inst_type == BGP_INSTANCE_TYPE_VRF)
 				continue;
 
-			bgp->router_id_zebra = router_id->u.prefix4;
-			if (!bgp->router_id_static.s_addr)
-				bgp_router_id_set(bgp, &router_id->u.prefix4);
+			if(addr)
+				bgp->router_id_zebra = *addr;
+			else
+				addr = &bgp->router_id_zebra;
+
+			if (!bgp->router_id_static.s_addr) {
+				if(bgp_router_id_valid(bgp) == false) {
```

Comment:
When the execution comes into this function as part of IP address delete (which was configured as router-id), then bgp_router_id_valid(bgp) should always return false and never returns true.  This is because as part of IP address delete, the address will be removed from address_hash and so this function returns false. This is only for delete case though. So can we modify the code accordingly ?